rfcs/45-stageless.md

```diff
@@ -0,0 +1,761 @@
+    - With the simplified run criteria, the stack is trivial to re-implement, but preferably a third-party plugin does it.
+    - The `apply_state_transition` system performs a single transition, doesn't do an endless cycle in one schedule pass.
+4. States no longer have an "on update".
+    - If a system must run on the same frame as the transition, it has to be in "on enter" and "on exit" (or your own DIY equivalent).
+
+## Rationale and alternatives
+
+### Why did you call them "sets" instead of "labels" or "categories"?
+
+The word "set" better describes what labeling actually means for scheduling.
+Sets aren't just an abstract tag, they have a "physical" location in the overall graph (a set is a "supernode" that encapsulates all the nodes inside it).
+Likewise, sets that intersect (share nodes) can't be made to depend on each other.
+
+### If users can configure system sets, can they mess with systems added by third-party plugins?
+
+Yes, somewhat, and it's intentional.
+
+Not being able to schedule plugin systems is a source of [significant user pain](https://github.com/bevyengine/bevy/issues/2160).
+Instead of the current status quo, we thought a better pattern would be a plugin exporting its top-level set label (at the very least) and the user being responsible for deciding where to put it.
+
+As it's now possible to define set hierarchies, plugin authors can decide exactly what they want to expose to users in their "public API" *and* still ensure any key logical invariants hold.
+Users will only be able schedule what the plugin author chooses to make public.
+Plugin authors could even make massive internal changes without breaking user apps (as long as the "public API" is relatively stable).
+
+### Why is there no "on-update" system set for states?
+
+Conceptually, a set occupies some position in a schedule.
+However, simply being in a state does not.
+Any system anywhere can be conditioned to only run if a certain state is active.
+If there was a builtin "on-update" set, it could only exist in one place.
+That, along with it being trivial to create a set with the condition that a state must be active, gave little reason to do it.
+
+```rust
+fn main() {
+    App::new()
+        /* ... */
+        .add_system(
+            my_system
+            // A state being active has no inherent position in the schedule.
+            // Any system anywhere can be told to only run if a certain state is active.
+            .in_set(CoreSet::PostUpdate)
+            // This helper generates a condition that returns `true` if the FSM is currently `GameState::Playing`.
+            .only_if(state_equals(GameState::Playing))
+        )
+        /* ... */
+        .run();
+    )
+}
+```
+
+## Unresolved questions
+
+### What sugar should we use for adding multiple systems at once?
+
+Convenience methods like `add_many` are important for reducing boilerplate. For these methods, we need to be able to refer to collections of systems and system sets.

```

Comment:
I personally dislike the naming here. `add_many` lacks specificity, and there are already many `App` extension traits that exist in this space, both within Bevy itself and in the ecosystem. IMO something like `add_systems` is unambiguous and is a marginal departure from the normal `add_system`.